### PR TITLE
Wii U Account Information App Support (attempt 2)

### PR DESCRIPTION
**Pull request by CaramelKat**
From #77, the branch had far too many merge conflicts and it was easier to just make a new one and close the old pr. 

This adds support for the Wii U account information app. So far the following information can be changed:
- Gender
- Region
- Timezone
- Email
- Server Environment
- Email Notifications
- Non-Nintendo Device Setting

This also requires this pr from [Inkay](https://github.com/PretendoNetwork/Inkay/pull/17), which has been merged already.

I have tested this on console locally, and it should have no impact on any other functionality on the account server. I think there is room to improve the UI, but that can be saved for the future. This included all the suggestions that had already been implemented from the previous pr. 

Screenshot:
image

### CODE REVIEW

**Comment by jonbarrow** [comment]
I'll take a deeper look later, but it should be noted that changing your region is NOT officially supported on NN

Because of this, changing it on our servers MIGHT cause some undefined behavior. We still need to support this, due to issues when making accounts on the website, but we should have a clear warning about this

**Comment by InternalLoss** [comment]
> I'll take a deeper look later, but it should be noted that changing your region is NOT officially supported on NN

This seems to be the **region** (as in the sub-region of the country), not the country itself. For example, United Kingdom has North East England, North West England, Wales, etc - these *are* all changeable on the system, just not the country of United Kingdom.

**Comment by CaramelKat** [comment]
> > I'll take a deeper look later, but it should be noted that changing your region is NOT officially supported on NN
> 
> This seems to be the **region** (as in the sub-region of the country), not the country itself. For example, United Kingdom has North East England, North West England, Wales, etc - these _are_ all changeable on the system, just not the country of United Kingdom.

Yep, that's correct. The settings that are here are the same that are present on the 3DS.

Also just so it's put out there, the layout isn't jank when you aren't a dev and only the two server environments are shown:

image

**Comment by jonbarrow** [comment]
The country is the only thing I'm talking about. The country is not changeable under normal circumstances. But we need to support changing it, since all accounts made on the website default to the US. This causes issues with mismatched regions in vWii at the very least

Hence why I said, this may cause undefined behavior and we need to add a warning

**Comment by CaramelKat** [comment]
As of now that field is read only, so if we want me to make to changeable with that warning I can totally do that here. We already have all the country codes from the regions anyway, so it'd be a quick change

**Comment by jonbarrow** [comment]
Sounds good. That definitely does need to be added. We've already confirmed that this mismatched region data prevents vWii users from going online, so being able to fix that is a must

**Comment by vonmillhausen** [comment]
I don't know if this would alter the scope, but being able to change date of birth somewhere would be cool as well; registering an account via the web site forces date of birth to be January 1st 1990, which currently it's impossible to change anywhere.

**Comment by DaniElectra** [comment]
I'm seeing that the changes for nnid/routes/people were dropped in this PR (specifically these: https://github.com/PretendoNetwork/account/pull/77/files#diff-8284dab686e24deaf626549249b36d236a61eb8fc0d5484d44f744bd35a7d52aR620-R625 ). Is this intentional? I think they should be added back either on this PR or on another

Aside from that, the [example.env](https://github.com/PretendoNetwork/account/blob/dev/example.env) and the [SETUP.md](https://github.com/PretendoNetwork/account/blob/dev/SETUP.md) should be updated to reflect the new environment variable

**Comment by CaramelKat** [comment]
Looks like I forgot to pull that over, I'll get that merged and the other files updated after I get off work here. If there's anything else obvious I missed just let me know. I had to manually cherry pick all these commits back out 🥲

File: src/assets/user-info-settings/index.css [hunk 1]
```diff
@@ -0,0 +1,298 @@
...
+	background-color: #FFF;
+	background-image: url("data:image/svg+xml,%3Csvg xmlns='http://www.w3.org/2000/svg' width='32' height='32' fill='%23673DB6' viewBox='0 0 256 256'%3E%3Cpath d='M213.66,101.66l-80,80a8,8,0,0,1-11.32,0l-80-80A8,8,0,0,1,53.66,90.34L128,164.69l74.34-74.35a8,8,0,0,1,11.32,11.32Z'%3E%3C/path%3E%3C/svg%3E");
+	background-size: 40px;
+	background-position: 440px center;
+	background-repeat: no-repeat;
+	color: #9D6FF3;
+	cursor: pointer;
+}
+
+
```
Comment (jonbarrow) [inline src/assets/user-info-settings/index.css:120]: Nitpick. 2 linebreaks

```suggestion

```

File: src/assets/user-info-settings/index.css [hunk 2]
```diff
@@ -0,0 +1,298 @@
...
+div.group > span {
+	color: #000;
+  }
+  div.group > svg {
```
Comment (jonbarrow) [inline src/assets/user-info-settings/index.css:124]: Nitpick. Spaces here

```suggestion
}
div.group > svg {
```

File: src/assets/user-info-settings/index.css [hunk 3]
```diff
@@ -0,0 +1,298 @@
...
+	fill: #9D6FF3;
+	border: none;
+	height: 40px;
+	width: 40px;
+	box-sizing: border-box;
+	user-select: none;
+	padding: 0px;
+	margin-bottom: -10px;
+	margin-left: 5px;
+  }
```
Comment (jonbarrow) [inline src/assets/user-info-settings/index.css:134]: Nitpick. Spaces here

```suggestion
}
```

File: src/assets/user-info-settings/index.css [hunk 4]
```diff
@@ -0,0 +1,298 @@
...
+	color: #45297A;
+	min-width: 200px;
+	box-sizing: content-box;
+	-webkit-box-sizing: content-box;
+	-webkit-box-align: center;
+	-webkit-box-pack: center;
+	min-height: 60px;
+	text-align: center;
+	font-size: 28px;
+	background: -webkit-gradient(linear, left top, left bottom, from(#FFF), color-stop(0.5, #FFF), color-stop(0.8, #F6F6F6), color-stop(0.95, #F5F5F5), to(#BBB))0 0;
```
Comment (jonbarrow) [inline src/assets/user-info-settings/index.css:52]: Nitpick. Missing space

```suggestion
	background: -webkit-gradient(linear, left top, left bottom, from(#FFF), color-stop(0.5, #FFF), color-stop(0.8, #F6F6F6), color-stop(0.95, #F5F5F5), to(#BBB)) 0 0;
```

File: src/assets/user-info-settings/index.css [hunk 5]
```diff
@@ -0,0 +1,298 @@
...
+div.radio label {
+	display: inline-block;
+	color: #6A6C75;
+	min-width: 200px;
+	box-sizing: content-box;
+	-webkit-box-sizing: content-box;
+	-webkit-box-align: center;
+	-webkit-box-pack: center;
+	text-align: center;
+	background: -webkit-gradient(linear, left top, left bottom, from(#FFF), color-stop(0.5, #FFF), color-stop(0.8, #F6F6F6), color-stop(0.95, #F5F5F5), to(#BBB))0 0;
```
Comment (jonbarrow) [inline src/assets/user-info-settings/index.css:250]: Nitpick. Missing space

```suggestion
	background: -webkit-gradient(linear, left top, left bottom, from(#FFF), color-stop(0.5, #FFF), color-stop(0.8, #F6F6F6), color-stop(0.95, #F5F5F5), to(#BBB)) 0 0;
```

File: src/assets/user-info-settings/index.css [hunk 6]
```diff
@@ -0,0 +1,298 @@
...
+
+.fixed-bottom-button,
+input[type="submit"] {
+	position: fixed;
+	bottom: 0;
+	right: 0;
+	height: 85px;
+	line-height: 100px;
+	padding: 0 40px 0 60px;
+	background: -webkit-gradient(linear, left top, left bottom, from(#FFF), color-stop(0.5, #FFF), to(#E6E6E6))0 0;
```
Comment (jonbarrow) [inline src/assets/user-info-settings/index.css:292]: Nitpick. Missing space

```suggestion
	background: -webkit-gradient(linear, left top, left bottom, from(#FFF), color-stop(0.5, #FFF), to(#E6E6E6)) 0 0;
```

File: src/services/nnas/index.ts [hunk 7]
```diff
@@ -15,17 +16,39 @@ import oauth from '@/services/nnas/routes/oauth';
 import people from '@/services/nnas/routes/people';
 import provider from '@/services/nnas/routes/provider';
 import support from '@/services/nnas/routes/support';
+import settings from '@/services/nnas/routes/account-settings';
 
 // * Router to handle the subdomain restriction
 const nnas = express.Router();
 
+// Static routes for the user information app
```
Comment (jonbarrow) [inline src/services/nnas/index.ts:24]: Nitpick. Make comment format match others

```suggestion
// * Static routes for the user information app
```

File: src/services/nnas/routes/account-settings.ts [hunk 8]
```diff
@@ -0,0 +1,215 @@
...
+import { HydratedPNIDDocument } from '@/types/mongoose/pnid';
+import { AccountSettings } from '@/types/services/nnas/account-settings';
+import { Token } from '@/types/common/token';
+import { RegionLanguages } from '@/types/services/nnas/region-languages';
+import { RegionTimezone, RegionTimezones } from '@/types/services/nnas/region-timezones';
+import { Country, Region } from '@/types/services/nnas/regions';
+import timezones from '@/services/nnas/timezones.json';
+import regionsList from '@/services/nnas/regions.json';
+
+const router: express.Router = express.Router();
```
Comment (jonbarrow) [inline src/services/nnas/routes/account-settings.ts:19]: We can remove this type def. We've started migrating to allowing TypeScript to infer types rather than always set them, unless we need to (reference commit https://github.com/PretendoNetwork/account/commit/cb10a1666812b994f58ebc2fe9a83f03901e6dcc)

This goes for every file here, I'll only comment it once though

```suggestion
const router = express.Router();
```

File: src/services/nnas/routes/account-settings.ts [hunk 9]
```diff
@@ -0,0 +1,215 @@
...
+ * Description: Serves the Nintendo Network ID Settings page for the Wii U
+ */
+router.get('/ui/profile', async function (request: express.Request, response: express.Response): Promise<void> {
+	const server: HydratedServerDocument | null = await getServerByClientID('3f3928cc6f780638d360f0485cef973f', config.server_environment);
+	const token: string | undefined = getValueFromHeaders(request.headers, 'x-nintendo-service-token');
+	if (!server || !token) {
+		response.sendStatus(504);
+		return;
+	}
+	const aes_key: string = server?.aes_key;
```
Comment (jonbarrow) [inline src/services/nnas/routes/account-settings.ts:44]: Spacing between these

```suggestion
	const token: string | undefined = getValueFromHeaders(request.headers, 'x-nintendo-service-token');

	if (!server || !token) {
		response.sendStatus(504);
		return;
	}

	const aes_key: string = server?.aes_key;
```

File: src/services/nnas/routes/account-settings.ts [hunk 10]
```diff
@@ -0,0 +1,215 @@
...
+			PNID,
+			regionTimezones,
+			face,
+			notice,
+			accountLevel,
+			regions: region ? region.regions: [],
+			regionsList
+		});
+	}
+	catch (error: any) {
```
Comment (jonbarrow) [inline src/services/nnas/routes/account-settings.ts:82]:
```suggestion
	} catch (error: any) {
```

File: src/services/nnas/routes/account-settings.ts [hunk 11]
```diff
@@ -0,0 +1,215 @@
...
+ * This is required because of the strict domain whitelist in the account settings app
+ * on the Wii U.
+ */
+router.get('/mii/:pid/:face', async function (request: express.Request, response: express.Response): Promise<void> {
+	if (!config.cdn.base_url) {
+		response.sendStatus(404);
+		return;
+	}
+	const miiImage: Buffer = await got(`${config.cdn.base_url}/mii/${request.params.pid}/${request.params.face}.png`).buffer();
+	response.set('Content-Type', 'image/png');
```
Comment (jonbarrow) [inline src/services/nnas/routes/account-settings.ts:101]:
```suggestion
	}

	const miiImage: Buffer = await got(`${config.cdn.base_url}/mii/${request.params.pid}/${request.params.face}.png`).buffer();

	response.set('Content-Type', 'image/png');
```

File: src/services/nnas/routes/account-settings.ts [hunk 12]
```diff
@@ -0,0 +1,215 @@
...
+});
+
+/**
+ * [POST]
+ * Description: Endpoint to update the PNID from the account settings app on the Wii
+ */
+router.post('/update', async function (request: express.Request, response: express.Response): Promise<void> {
+	const server: HydratedServerDocument | null = await getServerByClientID('3f3928cc6f780638d360f0485cef973f', config.server_environment);
+	const token: string | undefined = getValueFromHeaders(request.headers, 'x-nintendo-service-token');
+	if (!server || !token) {
```
Comment (jonbarrow) [inline src/services/nnas/routes/account-settings.ts:112]:
```suggestion
	const token: string | undefined = getValueFromHeaders(request.headers, 'x-nintendo-service-token');

	if (!server || !token) {
```

File: src/services/nnas/routes/account-settings.ts [hunk 13]
```diff
@@ -0,0 +1,215 @@
...
+		response.sendStatus(504);
+		return;
+	}
+
+	const aesKey: string = server?.aes_key;
+	const decryptedToken: Buffer = decryptToken(Buffer.from(token, 'base64'), aesKey);
+
+	const tokenContents: Token = unpackToken(decryptedToken);
```
Comment (jonbarrow) [inline src/services/nnas/routes/account-settings.ts:120]:
```suggestion
	const decryptedToken: Buffer = decryptToken(Buffer.from(token, 'base64'), aesKey);
	const tokenContents: Token = unpackToken(decryptedToken);
```

File: src/views/index.ejs [hunk 14]
```diff
@@ -0,0 +1,158 @@
...
+			var seen = false;
+			function warning() {
+				if(seen) return;
+				seen = true;
+				wiiuDialog.alert('Warning: Changing your country may cause unexpected results. You may need to re-link your PNID for the changes to apply. \n\nFor more information, see:\nhttps://pretendo.network/docs/errors/PAS-727-0001', 'I Understand');
+			}
+		</script>
+    </body>
+
+</html>
```
Comment (jonbarrow) [inline src/views/index.ejs:158]: This file has a big mix of spaces and tabs

File: src/views/index.ejs [hunk 15]
```diff
@@ -0,0 +1,158 @@
...
+        <script>
+                wiiuDialog.alert('<%= notice %>', 'OK');
+        </script>
+        <% } %>
+		<script>
+			var seen = false;
+			function warning() {
+				if(seen) return;
+				seen = true;
+				wiiuDialog.alert('Warning: Changing your country may cause unexpected results. You may need to re-link your PNID for the changes to apply. \n\nFor more information, see:\nhttps://pretendo.network/docs/errors/PAS-727-0001', 'I Understand');
```
Comment (jonbarrow) [inline src/views/index.ejs:153]: This link doesn't exist. Also we now own `preten.do` for short links, do you want to use that instead?

**Review by jonbarrow — COMMENTED** [review]
Mostly nitpicks

**Comment by CaramelKat** [comment]
Error codes have been added [here](https://github.com/PretendoNetwork/error-codes/pull/3).

File: src/cache.ts [hunk 16]
```diff
@@ -58,4 +58,4 @@ export async function getLocalCDNFile(name: string, encoding?: BufferEncoding):
 
 export async function setLocalCDNFile(name: string, value: Buffer): Promise<void> {
 	await setCachedFile(`local_cdn:${name}`, value);
-}
\ No newline at end of file
+}
```
Comment (jonbarrow) [inline src/cache.ts:61]: Only putting this here once, but why did you change the ling endings of so many files?

File: src/config-manager.ts [hunk 17]
```diff
@@ -148,6 +149,11 @@ if (!config.s3.secret) {
 	disabledFeatures.s3 = true;
 }
 
+if (!config.server_environment) {
+	LOG_WARN('Failed to find server environment. To change the environment, set the PN_ACT_CONFIG_SERVER_ENVIRONMENT environment variable');
```
Comment (jonbarrow) [inline src/config-manager.ts:153]: Should probably say it's using `prod` here

File: src/services/nnas/routes/account-settings.ts [hunk 18]
```diff
@@ -0,0 +1,228 @@
...
+				response.redirect(`/v1/account-settings/ui/profile?notice=${notice}`);
+				return;
+			}
+
+			if (environment === 'dev' && pnid.access_level < 3) {
+				response.status(400);
+				notice = 'Do not have permission to enter this environment';
+				response.redirect(`/v1/account-settings/ui/profile?notice=${notice}`);
+				return;
+			}
```
Comment (jonbarrow) [inline src/services/nnas/routes/account-settings.ts:199]: These access level checks remind me of the ones from the admin panel, which caused issues. I know there's basically no chance of that now, but better safe than sorry

File: src/cache.ts (hunk 16 above)
Comment (CaramelKat) [reply src/cache.ts:61]: Said this already in the discord, but I thought we had discussed that this was how we wanted to be going forward? I realize this was probably not the right place for that, but I updated my linter and felt like it was weird to only apply it to the couple files I was working on for this pr

Comment (DaniElectra) [reply src/cache.ts:61]: Maybe re can revert this commit [`68bcd1f` (#91)](https://github.com/PretendoNetwork/account/pull/91/commits/68bcd1f8a6559d85ddf267660e15e022d4d870a5) and apply it on a different PR?

Comment (CaramelKat) [reply src/cache.ts:61]: That works for me, I'll go ahead and revert it

**Review by jonbarrow — APPROVED** [review]

**Review by DaniElectra — APPROVED** [review]

**Comment by InternalLoss** [comment]
LGTM 💪

**Review by jonbarrow — APPROVED** [review]

**Comment by jonbarrow** [comment]
lgtm
